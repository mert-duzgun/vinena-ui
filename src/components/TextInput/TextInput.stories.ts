import type { Meta, StoryObj } from '@storybook/react';

import { TextInput } from './TextInput';

// More on how to set up stories at: https://storybook.js.org/docs/7.0/react/writing-stories/introduction
const meta = {
  title: 'TextInput',
  component: TextInput,
  tags: ['autodocs'],
} satisfies Meta<typeof TextInput>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/7.0/react/writing-stories/args
export const Default: Story = {
  args: {
    placeholder: 'Placeholder',
  },
};

export const WithLabel: Story = {
  args: {
    placeholder: 'Placeholder',
    label: 'Label',
  },
};

export const WithHelper: Story = {
  args: {
    placeholder: 'Placeholder',
    helperText: 'Helper text',
  },
};

export const Disabled: Story = {
  args: {
    placeholder: 'Placeholder',
    label: 'Label',
    helperText: 'Helper text',
    disabled: true,
  },
};

export const Error: Story = {
  args: {
    placeholder: 'Placeholder',
    label: 'Label',
    helperText: 'Helper text',
    error: true,
  },
};

export const WithLeftIcon: Story = {
  args: {
    placeholder: 'Placeholder',
    label: 'Label',
    helperText: 'Helper text',
    leftIcon: 'search',
  },
};

export const WithRightIcon: Story = {
  args: {
    placeholder: 'Placeholder',
    label: 'Label',
    helperText: 'Helper text',
    rightIcon: 'arrowRight',
  },
};

export const WithBothIcons: Story = {
  args: {
    placeholder: 'Placeholder',
    label: 'Label',
    helperText: 'Helper text',
    leftIcon: 'search',
    rightIcon: 'arrowRight',
  },
};
