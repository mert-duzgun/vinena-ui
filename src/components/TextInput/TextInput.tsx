import { useTheme } from '@emotion/react';
import { Box, Icon, IconName } from 'components';
import * as React from 'react';
import { PolymorphicComponentProp } from 'types';

import useStyles from './TextInput.styles';

export type TextInputProps<C extends React.ElementType> = PolymorphicComponentProp<
  C,
  {
    label?: string;
    id?: string;
    helperText?: string;
    disabled?: boolean;
    error?: boolean;
    leftIcon?: IconName;
    rightIcon?: IconName;
  }
>;

export const TextInput = <C extends React.ElementType = 'input'>({
  label,
  id,
  helperText,
  disabled = false,
  error = false,
  leftIcon,
  rightIcon,
  ...rest
}: TextInputProps<C>) => {
  const theme = useTheme();
  const css = useStyles(theme, { error, disabled, leftIcon });

  return (
    <Box css={css.root}>
      {label ? (
        <Box css={css.label} as={'label'} htmlFor={id}>
          {label}
        </Box>
      ) : null}
      <Box css={css.inputWrapper}>
        {leftIcon ? (
          <Icon testId="text-input-left-icon" name={leftIcon} css={css.leftIcon} />
        ) : null}
        <Box
          as={'input'}
          id={id}
          css={css.input}
          disabled={disabled}
          data-error={error}
          {...rest}
        />
        {rightIcon ? (
          <Icon testId="text-input-right-icon" name={rightIcon} css={css.rightIcon} />
        ) : null}
      </Box>
      {helperText ? <Box css={css.helperText}>{helperText}</Box> : null}
    </Box>
  );
};
