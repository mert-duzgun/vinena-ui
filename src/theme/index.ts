export * from './GlobalStyles';
export * from './theme';
export * from './utils';
