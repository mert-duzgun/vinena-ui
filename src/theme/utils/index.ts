export * from './px/px';
export * from './rem/rem';
