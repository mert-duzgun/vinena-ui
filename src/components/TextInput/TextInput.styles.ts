import { px, rem, VinenaTheme } from 'theme';

export interface TextInputStylesParams {
  error: boolean;
  disabled: boolean;
  leftIcon?: string;
}

export default function createStyles(theme: VinenaTheme, props: TextInputStylesParams) {
  return {
    root: {},
    label: {
      ...theme.fonts.p3,
      display: 'block',
      color: theme.colors.grayscale[600],
      marginBottom: rem(8),
    },
    inputWrapper: {
      position: 'relative',
    },
    input: {
      ...theme.fonts.p2,
      padding: `${rem(4)} ${rem(16)} ${rem(4)} ${props.leftIcon ? rem(50) : rem(16)}`,
      minHeight: rem(48),
      borderRadius: px(10),
      border: 'none',
      boxShadow: `inset 0 0 0 1px ${theme.colors.grayscale[500]}`,
      transition: 'box-shadow 0.2s ease-in-out',
      fontSize: rem(16),
      lineHeight: '1.5',
      color: theme.colors.grayscale[800],
      minWidth: rem(360),

      '&::placeholder': {
        opacity: 1,
        color: props.disabled ? theme.colors.grayscale[300] : theme.colors.grayscale[400],
      },
      '&:focus': {
        outline: 'none',
        boxShadow: `inset 0 0 0 2px ${theme.colors.primary[500]}`,
      },
      '&[data-error="true"]': {
        boxShadow: `inset 0 0 0 2px ${theme.colors.danger[500]}`,
      },
      '&:disabled': {
        backgroundColor: theme.colors.grayscale[100],
        boxShadow: `inset 0 0 0 1px ${theme.colors.grayscale[300]}`,
      },
    },
    leftIcon: {
      position: 'absolute',
      left: rem(16),
      top: '50%',
      transform: 'translateY(-50%)',
      fontSize: rem(24),
      color: props.disabled ? theme.colors.grayscale[300] : theme.colors.grayscale[400],
    },
    rightIcon: {
      position: 'absolute',
      right: rem(16),
      top: '50%',
      transform: 'translateY(-50%)',
      fontSize: rem(24),
      color: props.disabled ? theme.colors.grayscale[300] : theme.colors.grayscale[400],
    },
    helperText: {
      ...theme.fonts.p3,
      color: props.error ? theme.colors.danger[500] : theme.colors.grayscale[500],
      marginTop: rem(8),
    },
  } as const;
}
