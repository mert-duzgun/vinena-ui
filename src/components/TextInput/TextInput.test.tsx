import { ThemeProvider } from '@emotion/react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { DEFAULT_THEME } from 'theme';
import { vitest } from 'vitest';

import { TextInput } from './TextInput';

describe('Vinena/components/TextInput', () => {
  it('should match snapshot', async () => {
    const { container } = render(
      <ThemeProvider theme={DEFAULT_THEME}>
        <TextInput placeholder="Placeholder" />
      </ThemeProvider>
    );
    await expect(container.firstChild).toMatchSnapshot();
  });

  it('should render without errors', () => {
    render(
      <ThemeProvider theme={DEFAULT_THEME}>
        <TextInput />
      </ThemeProvider>
    );
  });

  it('should render the label and input element', () => {
    render(
      <ThemeProvider theme={DEFAULT_THEME}>
        <TextInput label="Test" id={'test'} />
      </ThemeProvider>
    );
    expect(screen.getByLabelText('Test')).toBeInTheDocument();
    expect(screen.getByRole('textbox')).toBeInTheDocument();
  });

  it('should render the helper text', () => {
    render(
      <ThemeProvider theme={DEFAULT_THEME}>
        <TextInput helperText="Test helper text" />
      </ThemeProvider>
    );
    expect(screen.getByText('Test helper text')).toBeInTheDocument();
  });

  it('should disable the input element when disabled prop is passed', () => {
    render(
      <ThemeProvider theme={DEFAULT_THEME}>
        <TextInput disabled />
      </ThemeProvider>
    );
    expect(screen.getByRole('textbox')).toBeDisabled();
  });

  it('should display error colors when error prop is passed', () => {
    render(
      <ThemeProvider theme={DEFAULT_THEME}>
        <TextInput error helperText="Test helper text" />
      </ThemeProvider>
    );
    expect(screen.getByRole('textbox')).toHaveStyle('box-shadow: inset 0 0 0 2px #E12D39;');
    expect(screen.getByText('Test helper text')).toHaveStyle('color: #E12D39;');
  });

  it('should render the left icon', () => {
    render(
      <ThemeProvider theme={DEFAULT_THEME}>
        <TextInput leftIcon="check" />
      </ThemeProvider>
    );

    expect(screen.getByTestId('text-input-left-icon')).toBeInTheDocument();
  });

  it('should render the right icon', () => {
    render(
      <ThemeProvider theme={DEFAULT_THEME}>
        <TextInput rightIcon="check" />
      </ThemeProvider>
    );
    expect(screen.getByTestId('text-input-right-icon')).toBeInTheDocument();
  });

  it('should render both left and right icons when both props are passed', () => {
    render(
      <ThemeProvider theme={DEFAULT_THEME}>
        <TextInput leftIcon="check" rightIcon="check" />
      </ThemeProvider>
    );
    expect(screen.getAllByTestId(/icon/i).length).toEqual(2);
  });

  it('should handle when no props are passed', () => {
    render(
      <ThemeProvider theme={DEFAULT_THEME}>
        <TextInput />
      </ThemeProvider>
    );
    expect(screen.queryByLabelText('')).not.toBeInTheDocument();
    expect(screen.getByRole('textbox')).toBeInTheDocument();
    expect(screen.queryByTestId(/icon/i)).not.toBeInTheDocument();
    expect(screen.queryByText(/helper/i)).not.toBeInTheDocument();
  });

  it('should call onChange handler when input is typed', async () => {
    const mockOnChange = vitest.fn();
    render(
      <ThemeProvider theme={DEFAULT_THEME}>
        <TextInput onChange={mockOnChange} />
      </ThemeProvider>
    );
    const input = screen.getByRole('textbox');
    await userEvent.type(input, 'test');
    expect(mockOnChange).toHaveBeenCalledTimes(4);
  });
});
