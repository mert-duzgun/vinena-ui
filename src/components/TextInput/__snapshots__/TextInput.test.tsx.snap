// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Vinena/components/TextInput > should match snapshot 1`] = `
<div
  class="css-167j6s7-TextInput"
>
  <div
    class="css-7117jv-TextInput"
  >
    <input
      class="css-rrm5su-TextInput"
      data-error="false"
      placeholder="Placeholder"
    />
  </div>
</div>
`;
